### PR TITLE
Type annotating tmt/steps/finish/__init__.py

### PR DESCRIPTION
**Pull request by guy9050**

### CODE REVIEW

File: tmt/steps/finish/__init__.py [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
 
 import click
 import fmf
 
 import tmt
+from tmt.steps import Step
 
 
-class Finish(tmt.steps.Step):
+class Finish(Step):
```
Comment (thrix) [inline tmt/steps/finish/__init__.py:11]: @guy9050 I would keep it as `tmt.steps.Step`, it is what other modules use, seems no benefit and breaks this consistency

**Review by thrix — CHANGES_REQUESTED** [review]

**Comment by guy9050** [comment]
Looks like there is inconsistency between `def go(self, guest: Guest) -> None:` (line 56)
and how it's defined in the parent class [here](https://github.com/teemtee/tmt/blob/main/tmt/steps/__init__.py#L561): `def go(self, *args: Any, **kwargs: Any) -> None:`

What should be the right definition then?

**Comment by thrix** [comment]
> Looks like there is inconsistency between `def go(self, guest: Guest) -> None:` (line 56) and how it's defined in the parent class [here](https://github.com/teemtee/tmt/blob/main/tmt/steps/__init__.py#L561): `def go(self, *args: Any, **kwargs: Any) -> None:`
> 
> Mypy complains about that: `tmt/steps/finish/shell.py:56: error: Signature of "go" incompatible with supertype "Plugin" `
> 
> What should be the right definition then?

I would say the args and kwargs should be added to the parent function

**Comment by thrix** [comment]
> > Looks like there is inconsistency between `def go(self, guest: Guest) -> None:` (line 56) and how it's defined in the parent class [here](https://github.com/teemtee/tmt/blob/main/tmt/steps/__init__.py#L561): `def go(self, *args: Any, **kwargs: Any) -> None:`
> > Mypy complains about that: `tmt/steps/finish/shell.py:56: error: Signature of "go" incompatible with supertype "Plugin" `
> > What should be the right definition then?
> 
> I would say the args and kwargs should be added to the parent function

We discussed that we will need to discuss this with @happz once back, added to discussion list in https://github.com/teemtee/tmt/issues/1182

**Comment by guy9050** [comment]
blocked until  `go()` is moved out of `Phase` to being implemented in subclasses. @happz is working on that.

**Comment by thrix** [comment]
@guy9050 this is now unblocked, please rebase

**Comment by thrix** [comment]
go problem - should be solved, rebase
the listify problem was solved in other MR

**Comment by thrix** [comment]
@happz @ukulekek @janhavlin I would maybe go with this as it is ... We will revisit ignores anyway .,.. if it passes ...

**Comment by thrix** [comment]
@guy9050 pls do not use `wip` in the PR summary, rather please use draft PRs. I removed the obscure `wip`

**Review by thrix — APPROVED** [review]

File: tmt/steps/finish/__init__.py [hunk 2]
```diff
@@ -22,8 +21,9 @@ class Finish(tmt.steps.Step):
     steps failed (for example when the environment preparation was not
     successful) so that provisioned systems are not kept running.
     """
+    data: List[StepData]
```
Comment (thrix) [inline tmt/steps/finish/__init__.py:24]: @guy9050 why is this needed, data is defined in `tmt.steps.Step` and it has `Union[StepData, List[StepData]]` type?

File: tmt/steps/finish/shell.py [hunk 3]
```diff
@@ -33,11 +33,11 @@ class FinishShell(tmt.steps.finish.FinishPlugin):  # type: ignore[misc]
     @classmethod
     def options(cls, how: Optional[str] = None) -> Any:
         """ Finish command line options """
-        return [
+        return cast(List[tmt.options.ClickOptionDecoratorType], [
```
Comment (thrix) [inline tmt/steps/finish/shell.py:36]: please look @ https://github.com/teemtee/tmt/pull/1418/files#diff-869f5ba1150a929f89989f134e4591435b5c0a66e2cf553a046357239a135b7eR57

and similar how we agreed to type this in other modules

**Review by thrix — CHANGES_REQUESTED** [review]

File: tmt/steps/finish/__init__.py (hunk 2 above)
Comment (happz) [reply tmt/steps/finish/__init__.py:24]: It's something I try to address in #1384, `self.data` may be either a single object or a list of objects - that's what annotations say, but in reality, it's always a list -, and often a particular plugin treats `self.data` as the list only. A developer focusing on that particular plugin cannot resolve the original issue, and adding `List[StepData]` makes annotating the plugin easier.

@janhavlin mentioned these in #1384, and I'll resolve them in my patch (once I'm done with failing tests...).

File: tmt/steps/finish/shell.py (hunk 3 above)
Comment (guy9050) [reply tmt/steps/finish/shell.py]: Changed. Thanks.

File: tmt/steps/finish/__init__.py (hunk 2 above)
Comment (guy9050) [reply tmt/steps/finish/__init__.py:24]: @thrix does that answer your question?

File: tmt/steps/finish/shell.py [hunk 4]
```diff
@@ -31,13 +31,16 @@ class FinishShell(tmt.steps.finish.FinishPlugin):  # type: ignore[misc]
     _keys = ["script"]
 
     @classmethod
-    def options(cls, how: Optional[str] = None) -> Any:
+    def options(cls, how: Optional[str] = None) -> List[tmt.options.ClickOptionDecoratorType]:
         """ Finish command line options """
-        return [
+        # TODO: revisit this cast once super class is typed
```
Comment (janhavlin) [inline tmt/steps/finish/shell.py:36]: This comment is not needed anymore?

File: tmt/steps/finish/__init__.py (hunk 2 above)
Comment (thrix) [reply tmt/steps/finish/__init__.py:24]: yep, ty!

**Review by thrix — APPROVED** [review]
LGTM

File: tmt/utils.py [hunk 5]
```diff
@@ -870,7 +871,7 @@ def listify(
     if isinstance(data, dict):
         for key in keys or data:
             if key in data:
-                data[key] = listify(data[key], split=split)
+                data[key] = listify(data[key], split=split)  # type: ignore
```
Comment (janhavlin) [inline tmt/utils.py:874]: I have 2 notes regarding this function.

First, it doesn't make much sense to have return type `StepData` alongside `Dict` and `List` due to the character of the function (how it transforms the input data). If you feed it `StepData`, it will change it into something that is not `StepData` type anymore:

E.g. code `print(tmt.utils.listify(tmt.steps.StepData(name='hello', order=None)))` prints out: `{'name': ['hello'], 'order': [None]}` which is clearly not `StepData` anymore due to the nested lists. The only way to get `StepData` as an output seems to be by inputting an empty dictionary but why would one use this function then? :)

Second, if we convert `TypedDict` to `Dict`, we can get rid of the mypy complaint that made us insert the `type: ignore`. Converting to `Dict` IMO makes sense due to 1).

My proposal is to change the function to this:
```
def listify(
        data: Union[Tuple[Any, ...], List[Any], str, Dict[Any, Any], 'tmt.steps.StepData'],
        split: bool = False,
        keys: Optional[List[str]] = None) -> Union[List[Any], Dict[Any, Any]]:
    """
    Ensure that variable is a list, convert if necessary

    For dictionaries check all items or only those with provided keys.
    Also split strings on white-space/comma if split=True.
    """
    separator = re.compile(r'[\s,]+')
    if isinstance(data, tuple):
        data = list(data)
    if isinstance(data, list):
        return fmf.utils.split(data, separator) if split else data
    if isinstance(data, str):
        return fmf.utils.split(data, separator) if split else [data]
    if isinstance(data, dict):
        data_dict = dict(data)
        for key in keys or data_dict:
            if key in data_dict:
                data_dict[key] = listify(data_dict[key], split=split)
        return data_dict
    return [data]
```

File: tmt/steps/finish/shell.py (hunk 4 above)
Comment (guy9050) [reply tmt/steps/finish/shell.py:36]: ty

File: tmt/utils.py (hunk 5 above)
Comment (guy9050) [reply tmt/utils.py:874]: I totally get your point about why providing `StepData` as another accepted type for `data` in `listify(...)` may not make much sense, and I agree with it. But since we discussed this PR in the team gathering with @happz and @thrix and agreed on adding 'tmt.steps.StepData' as another accepted type for `data` in `listify`, I'd like to hear their thoughts before I make further changes.

Comment (janhavlin) [reply tmt/utils.py:874]: It makes sense to **accept** `StepData`, we already discussed it on team meetings I believe. But it doesn't make much sense to **return** `StepData`. See how it transforms and outputs the data - they do not resemble `StepData` at all.

But perhaps this discussion isn't so important because `StepData` is supposed to be refactored anyhow? IMO my change is still worth it because it makes the function look more clear and helps get rid of the `type: ignore`. :)

Comment (happz) [reply tmt/utils.py:874]: This part will probably change very Soon (TM), with `StepData` refactoring, so I wouldn't invest too much time into this. Accepting `StepData` is correct, returning them isn't, because `listify()` is never going to return a `StepData` instance. I would add `# type: ignore` comment, probably with more fitting error code, and a big `TODO: should go away with StepData becoming a dataclass` comment, and call it a day :)

Comment (thrix) [reply tmt/utils.py:874]: +1
@guy9050 pls make the change so we can merge this, it is blocking #1300

Comment (thrix) [reply tmt/utils.py:874]: and also rebase

Comment (guy9050) [reply tmt/utils.py:902]: Done

**Review by janhavlin — APPROVED** [review]

File: tmt/steps/finish/shell.py (hunk 4 above)
Comment (guy9050) [inline tmt/steps/finish/shell.py:36]:
```suggestion
```
